stdlib/public/core/Diffing.swift

```diff
@@ -0,0 +1,805 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2015 - 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+// MARK: Diff application to RangeReplaceableCollection
+
+@available(macOS 9999, iOS 9999, tvOS 9999, watchOS 9999, *) // FIXME(availability-5.1)
+extension CollectionDifference {
+  fileprivate func _fastEnumeratedApply(
+    _ consume: (Change) -> Void
+  ) {
+    let totalRemoves = removals.count
+    let totalInserts = insertions.count
+    var enumeratedRemoves = 0
+    var enumeratedInserts = 0
+
+    while enumeratedRemoves < totalRemoves || enumeratedInserts < totalInserts {
+      let change: Change
+      if enumeratedRemoves < removals.count && enumeratedInserts < insertions.count {
+        let removeOffset = removals[enumeratedRemoves]._offset
+        let insertOffset = insertions[enumeratedInserts]._offset
+        if removeOffset - enumeratedRemoves <= insertOffset - enumeratedInserts {
+          change = removals[enumeratedRemoves]
+        } else {
+          change = insertions[enumeratedInserts]
+        }
+      } else if enumeratedRemoves < totalRemoves {
+        change = removals[enumeratedRemoves]
+      } else if enumeratedInserts < totalInserts {
+        change = insertions[enumeratedInserts]
+      } else {
+        // Not reached, loop should have exited.
+        preconditionFailure()
+      }
+
+      consume(change)
+
+      switch change {
+      case .remove(_, _, _):
+        enumeratedRemoves += 1
+      case .insert(_, _, _):
+        enumeratedInserts += 1
+      }
+    }
+  }
+}
+
+extension RangeReplaceableCollection {
+  /// Applies a difference to a collection.
+  ///
+  /// - Parameter difference: The difference to be applied.
+  ///
+  /// - Returns: An instance representing the state of the receiver with the
+  ///   difference applied, or `nil` if the difference is incompatible with
+  ///   the receiver's state.
+  ///
+  /// - Complexity: O(*n* + *c*), where *n* is `self.count` and *c* is the
+  ///   number of changes contained by the parameter.
+  @available(macOS 9999, iOS 9999, tvOS 9999, watchOS 9999, *) // FIXME(availability-5.1)
+  public func applying(_ difference: CollectionDifference<Element>) -> Self? {
+    var result = Self()
+    var enumeratedRemoves = 0
+    var enumeratedInserts = 0
+    var enumeratedOriginals = 0
+    var currentIndex = self.startIndex
+
+    func append(
+      into target: inout Self,
+      contentsOf source: Self,
+      from index: inout Self.Index, count: Int) {
```

Comment:
The convention is to put a closing paren and the opening curly on the new line unindented after the parameter list.